cmd/check-markdown-rs/src/main.rs

```diff
@@ -0,0 +1,318 @@
+            }
+            _ => {}
+        }
+    }
+
+    if errors.is_empty() {
+        Ok(())
+    } else {
+        Err(errors)
+    }
+}
+
+// Recursive function to find all heading nodes in the AST
+fn find_heading_ids<'a>(node: &'a Node<'a, RefCell<Ast>>, heading_ids: &mut HashSet<String>) {
+    match &node.data.borrow().value {
+        NodeValue::Heading(heading) => {
+            // Use one of the available fields for NodeHeading
+            let level = heading.level;
+            // Do something with level...
+        }
+        _ => {}
+    }
+
+    for child in node.children() {
+        find_heading_ids(&child, heading_ids);
+    }
+}
+
+// Recursive function to find all link nodes in the AST
+fn find_link_nodes<'a>(
+    node: &'a Node<'a, RefCell<Ast>>,
+    link_nodes: &mut Vec<&'a Node<'a, RefCell<Ast>>>,
+) {
+    match &node.data.borrow().value {
+        NodeValue::Link(_) => link_nodes.push(node),
+        _ => {}
+    }
+
+    for child in node.children() {
+        find_link_nodes(&child, link_nodes);
+    }
+}
+
+// Function to gather and print document statistics
+fn print_document_statistics<'a>(root: &'a Node<'a, RefCell<Ast>>) -> HashMap<&'a str, usize> {
+    let mut statistics: HashMap<&str, usize> = HashMap::new();
+    gather_document_statistics(root, &mut statistics);
+
+    statistics
+}
+
+// Recursive function to gather document statistics
+fn gather_document_statistics<'a>(
+    node: &'a Node<'a, RefCell<Ast>>,
+    statistics: &mut HashMap<&'a str, usize>,
+) {
+    match &node.data.borrow().value {
+        NodeValue::Heading(_) => {
+            *statistics.entry("headings").or_insert(0) += 1;
+        }
+        NodeValue::Text(text) => {
+            let word_count = String::from_utf8_lossy(text.as_bytes())
+                .split_whitespace()
+                .count();
+            *statistics.entry("words").or_insert(0) += word_count;
+        }
+        NodeValue::Link(_) => {
+            *statistics.entry("links").or_insert(0) += 1;
+        }
+        _ => {}
+    }
+
+    for child in node.children() {
+        gather_document_statistics(&child, statistics);
+    }
+}
+
+fn generate_output(
+    toc: &str,
+    link_validation_result: &Result<(), Vec<String>>,
+    structure_validation_result: &Result<(), String>,
+    document_statistics: &HashMap<&str, usize>,
+) {
+    println!("Final Output:");
+    println!("\nDocument Structure Validation:");
+    match structure_validation_result {
+        Ok(_) => println!("  The document structure is valid."),
+        Err(err) => println!("  Error: {}", err),
+    }
+
+    println!("\nTable of Contents:");
+    println!("{}", toc);
+
+    println!("\nLink Validation:");
+    match link_validation_result {
+        Ok(_) => println!("  All links are valid."),
+        Err(errors) => {
+            println!("  Found {} invalid links:", errors.len());
+            for error in errors {
+                println!("    {}", error);
+            }
```

Comment:
I like the way you've collected all the errors together here and are displaying them at the end. Conventionally, errors should be displayed to standard error stream, so can you change the `println!()` calls to `eprintln!()` and add an `"ERROR: "` prefix to the call in the for loop?